summary: Test httputil/retry.go ShouldRetryError(err)/NoNetwork(err)

# ubuntu-core: no go-compiler
# ubuntu-14.04: no nsenter
systems: [-ubuntu-core-*, -ubuntu-14.04-*]

prepare: |
    go build -o detect-retry ./detect-retry.go

restore: |
    ip netns delete testns || true
    umount /run/netns || true

debug: |
   cat output.txt || true
   cat stderr || true

execute: |
    echo "Add totally unconnected network namespace"
    ip netns add testns

    # Without having "lo" up the ipv6 test will retrun EADDRNOTAVAIL which
    # is misleading, it just means that there is no ipv6 addresses in the
    # namespace, adding localhost fixes this.
    echo "but make sure we have localhost available so that ipv6 works"
    ip netns exec testns ip link set dev lo up

    SNAPD_DEBUG=1 nsenter --net=/var/run/netns/testns ./detect-retry http://www.ubuntu.com > output.txt 2>stderr
    # XXX: ShouldRetryError is slightly misbehaving, if we know we don't have
    #      a network connection we should not retry. However we keep it as
    #      it is for now to be sure we don't add regressions.
    MATCH "ShouldRetryError: true" < output.txt
    MATCH "NoNetwork: true" < output.txt
    # be paranoid and look at the low-level go error as well
    MATCH "(Temporary failure in name resolution|network is unreachable)" < stderr

    echo "Now without DNS resolver with ipv4"
    UBUNTU_IP="$(getent ahostsv4 www.ubuntu.com|awk '{print $1 }' | head -n1)"
    SNAPD_DEBUG=1 nsenter --net=/var/run/netns/testns ./detect-retry "http://$UBUNTU_IP" > output.txt 2>stderr
    # XXX: ShouldRetryError is slightly misbehaving, see comment above
    MATCH "ShouldRetryError: true" < output.txt
    MATCH "NoNetwork: true" < output.txt
    # be paranoid and look at the low-level go error as well
    MATCH "network is unreachable" < stderr

    echo "Now without DNS resolver for ipv6"
    UBUNTU_IP="$(getent ahostsv6 www.ubuntu.com|awk '{print $1 }' | head -n1)"
    SNAPD_DEBUG=1 nsenter --net=/var/run/netns/testns ./detect-retry "http://[$UBUNTU_IP]" > output.txt 2>stderr
    # XXX: ShouldRetryError is slightly misbehaving, see comment above
    MATCH "ShouldRetryError: true" < output.txt
    MATCH "NoNetwork: true" < output.txt
    # be paranoid and look at the low-level go error as well
    MATCH "network is unreachable" < stderr
